contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,860 @@
+                underlyingAssets
+            );
+        }
+
+        // burn remaining OETHb
+        _burnOETHbOnTheContract();
+        positionManager.approve(address(clGauge), tokenId);
+        clGauge.deposit(tokenId);
+    }
+
+    /**
+     * @dev Check that the liquidity in the pool is withing the expected WETH to OETHb ratio
+     */
+    function _checkLiquidityWithinExpectedShare() internal {
+        (uint256 WETHPositionBalance, uint256 OETHbPositionBalance) = getPositionPrincipal();
+        require(WETHPositionBalance + OETHbPositionBalance > 0, "Can not withdraw full position");
+
+        uint160 currentPrice = getPoolX96Price();
+        // check we are in inspected tick range
+        require(currentPrice >= sqrtRatioX96Tick0 && currentPrice <= sqrtRatioX96Tick1, "Not in expected tick range");
+
+        uint256 currentWethShare = 0;
+        if (WETHPositionBalance != 0) {
+            currentWethShare = WETHPositionBalance.divPrecisely(WETHPositionBalance + OETHbPositionBalance);
+        }
+
+        if (currentWethShare < poolWethShareVarianceAllowed || // uint256 mustn't go below 0  
+            poolWethShare < currentWethShare - poolWethShareVarianceAllowed ||
+            poolWethShare > currentWethShare + poolWethShareVarianceAllowed) {
+
+            revert PoolRebalanceOutOfBounds(
+                currentWethShare,
+                poolWethShare,
+                WETHPositionBalance,
+                OETHbPositionBalance
+            );
+        } else {
+            emit PoolRebalanced(
+                currentWethShare,
+                poolWethShare,
+                WETHPositionBalance,
+                OETHbPositionBalance
+            );
+        }
+    }
+
+    /**
+     * Burns any lingering OETHb tokens still remaining on the contract
+     */
+    function _burnOETHbOnTheContract() internal {
+        uint256 OETHbBalance = IERC20(OETHb).balanceOf(address(this));
+        IVault(vaultAddress).burnForStrategy(OETHbBalance);
+    }
+
+    function updateUnderlyingAssets() internal {
+        if (tokenId == 0) {
+            underlyingAssets = 0;
+        } else {
+            (uint128 liquidity,,) = _getPositionInfo();
+
+            /**
+             * Our net value represent the smallest amount of tokens we are able to extract from the position
+             * given our liquidity.
+             *
+             * The least amount of tokens extraditable from the position is where the active trading price is
+             * at the ticker 0 meaning the pool is offering 1:1 trades between WETH & OETHb. At that moment the pool 
+             * consists completely of OETHb and no WETH.
+             * 
+             * The more swaps from WETH -> OETHb happen on the pool the more the price starts to move towards the -1
+             * ticker making OETHb (priced in WETH) more expensive.
+             */
+            (uint256 wethAmount, uint256 OETHbAmount) = helper.getAmountsForLiquidity(
```

Comment:
Optional, but we could do `getAmountsForLiquidity` once in the constructor, then get a constant multiplier out of it that we could use for each update.

Reply:
Liquidity is the property that is constantly changing. So yes we could getAmounts for a unit of liquidity and in this function multiply it with however many units of liquidity are present when this function is called.

I am worried though that this gas optimisation would increase complexity and reduce readability. And, at least on base network where gas is cheap, I don't think it is worth it?

Reply:
We can leave it.